Added epoch state cache configuration

rolfyone opened the pull request:
Now that attestations aren't being validated against states so often, we need a cache closer aligned to what we're doing when we need to regenerate - basically we want to avoid regeneration of state if at all possible.

One epoch transition the finalized and sometimes justified state get read, and were causing regenerations, so adding a cache to handle storing those, but leaving it size 0 by default for now.

The idea would be to have the current state cache reduced significantly in size, potentially to about half an epoch or so, and have an epoch cache with a maximum size over 2, and could probably be below 10 by default.

This would mean the vast majority of the time, no state regeneration is required, and our states in memory can drop to a much lower number than currently stored in memory (32 * 5 max by default currently).

Tested in minimal network, needs to be deployed to a more complex setup to see if it behaves well in a real environment.

rolfyone committed:
Added epoch state cache configuration

Now that attestations aren't being validated against states so often, we need a cache closer aligned to what we're doing when we need to regenerate - basically we want to avoid regeneration of state if at all possible.

One epoch transition the finalized and sometimes justified state get read, and were causing regenerations, so adding a cache to handle storing those, but leaving it size 0 by default for now.

The idea would be to have the current state cache reduced significantly in size, potentially to about half an epoch or so, and have an epoch cache with a maximum size over 2, and could probably be below 10 by default.

This would mean the vast majority of the time, no state regeneration is required, and our states in memory can drop to a much lower number than currently stored in memory (32 * 5 max by default currently).

Tested in minimal network, needs to be deployed to a more complex setup to see if it behaves well in a real environment.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

ajsutton commented on storage/src/main/java/tech/pegasys/teku/storage/store/Store.java:
```diff
@@ -609,14 +640,43 @@ private SafeFuture<Optional<StateAndBlockSummary>> getOrRegenerateBlockAndState(
+              .filter(
+                  stateAndBlockSummary ->
+                      stateAndBlockSummary.getBlockSummary().getRoot().equals(blockRoot))
+              .findFirst();
```
Might be worth making `maybeEpochStates` a map keyed by block root to make lookups faster?

ajsutton commented on storage/src/main/java/tech/pegasys/teku/storage/store/Store.java:
```diff
@@ -609,14 +640,43 @@ private SafeFuture<Optional<StateAndBlockSummary>> getOrRegenerateBlockAndState(
+      final UInt64 slot = maybeStateAndBlockSummary.get().getSlot();
+      final int slotsPerEpoch =
+          spec.atSlot(maybeStateAndBlockSummary.get().getSlot()).getConfig().getSlotsPerEpoch();
+      if (slot.mod(slotsPerEpoch).isZero()) {
```
This is going to miss epochs where the first slot is empty. I suspect the key question is whether we'd want to cache the first block from the epoch or if it's better to make this a cache of checkpoint states where the state is from the last block of the previous epoch with the state transition applied. The nice thing about using that checkpoint state approach is that happens to be what we pre-cache prior to the epoch transition so we'd always generate that state and then it can be used to apply the first block of the epoch without redoing the transition regardless of how many empty slots are present at the start of the epoch. It can also be used as-is for any duty calculations that needs a state from that epoch.

But it probably depends which lookups are trying to hit this cache - it wouldn't really work for getting a state by block root in that case which I think is more what you're targeting for getting the justified/finalized states.

rolfyone replied:
given how the 'hot state persistence' is done, i might be able to more closely integrate with that - will take another pass at it...

ajsutton commented on storage/src/main/java/tech/pegasys/teku/storage/store/Store.java:
```diff
@@ -747,4 +807,24 @@ private void putBlock(final SignedBeaconBlock block) {
+  void removeStateAndBlock(final Bytes32 root) {
+    blocks.remove(root);
+    states.remove(root);
+  }
```
This probably should remove from epoch states too which would work well if it's keyed by block root (not so well if it winds up being the checkpoint state cache I mention above...). In long periods of non-finalisation you may have some relatively long chains invalidated when finalisation invalidates a fork.

rolfyone replied:
that was my initial intent, it would have been so simple...
pruned hot blocks listed the finalized block so it deleted the associated state, so i handled the epochs separately initially so that it remains...

ajsutton reviewed (commented):
Other thing to consider here is that this epoch state cache probably should be setup to ensure it works well with the logic to persist hot states periodically which is also looking for states at the epoch boundary. I'm not sure if that's dependent on what's in the cache to be able to persist it or not.

rolfyone committed:
Merge remote-tracking branch 'upstream/master' into epoch-state-cache

rolfyone committed:
changed to using map and setting justified and finalized in storeTransactionUpdates.

This should address all of the review feedback, and the cleanups I made while handling less healthy networking conditions.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
store last state before epoch transition when the first slot of the epoch has no block

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
clarify comment

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
rename new metric to be more consistent with the other state cache metrics.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

tbenr commented on storage/src/main/java/tech/pegasys/teku/storage/store/Store.java:
```diff
@@ -609,14 +637,46 @@ private SafeFuture<Optional<StateAndBlockSummary>> getOrRegenerateBlockAndState(
                     : EmptyStoreResults.EMPTY_STATE_AND_BLOCK_SUMMARY_FUTURE);
   }
 
+  private void cacheIfEpochState(Optional<StateAndBlockSummary> maybeStateAndBlockSummary) {
```
`final`

tbenr commented on storage/src/main/java/tech/pegasys/teku/storage/store/Store.java:
```diff
@@ -575,7 +599,11 @@ VoteTracker getVote(UInt64 validatorIndex) {
     return getOrRegenerateBlockAndState(blockRoot)
-        .thenApply(res -> res.map(StateAndBlockSummary::getState));
+        .thenApply(
+            res -> {
```
I'd rename to `maybeStateAndBlockSummary` for readability

tbenr replied:
do we break anything if we move `cacheIfEpochState` from here to directly in `getOrRegenerateBlockAndState:638` ?
seems like more consistent and we won't try to put result in `epochState` cache when there is a cache hit from `epochState` itself.

rolfyone replied:
wish I had left some trace in - should be ok, i'll instrument it again... good pick up maybe...

rolfyone replied:
the problem is the other getBlockState, will refactor but its currently separate code paths.

tbenr commented on storage/src/main/java/tech/pegasys/teku/storage/store/Store.java:
```diff
@@ -609,14 +637,46 @@ private SafeFuture<Optional<StateAndBlockSummary>> getOrRegenerateBlockAndState(
 
+  private void cacheIfEpochState(Optional<StateAndBlockSummary> maybeStateAndBlockSummary) {
+    if (maybeStateAndBlockSummary.isPresent() && maybeEpochStates.isPresent()) {
+      final StateAndBlockSummary summary = maybeStateAndBlockSummary.get();
```
I'd name it as `stateAndBlockSummary`

tbenr commented on storage/src/main/java/tech/pegasys/teku/storage/store/Store.java:
```diff
@@ -609,14 +637,46 @@ private SafeFuture<Optional<StateAndBlockSummary>> getOrRegenerateBlockAndState(
+      if (!slot.mod(spec.getSlotsPerEpoch(slot)).isZero()) {
+        // pre-epoch transition state
+        // This will be referenced during epoch transition if the first slot of the epoch is empty
+        final Optional<StateAndBlockSummary> maybeParent =
```
`maybeParentStateAndBlockSummary`

tbenr commented on storage/src/main/java/tech/pegasys/teku/storage/store/Store.java:
```diff
@@ -609,14 +637,46 @@ private SafeFuture<Optional<StateAndBlockSummary>> getOrRegenerateBlockAndState(
+        final Optional<StateAndBlockSummary> maybeParent =
+            states.getIfAvailable(summary.getParentRoot());
+        maybeParent.ifPresent(
+            stateAndBlockSummary -> epochStates.put(summary.getParentRoot(), stateAndBlockSummary));
```
`parentStateAndBlockSummary`

tbenr reviewed (commented):
for now just commenting some nits. I want to run some tests too before approving

tbenr commented on storage/src/main/java/tech/pegasys/teku/storage/store/StoreTransactionUpdates.java:
```diff
@@ -118,16 +118,24 @@ public void applyToStore(final Store store, final UpdateResult updateResult) {
     finalizedChainData.ifPresent(
-        finalizedData -> store.finalizedAnchor = finalizedData.getLatestFinalized());
+        finalizedData -> {
+          store.maybeEpochStates.ifPresent(
```
can we add some comments to cover the reasoning? or move this in a method with a meaningful name

rolfyone replied:
i pretty much rewrote all of this should be clearer now

tbenr commented on storage/src/main/java/tech/pegasys/teku/storage/store/Store.java:
```diff
@@ -609,14 +637,46 @@ private SafeFuture<Optional<StateAndBlockSummary>> getOrRegenerateBlockAndState(
     }
+
+    // is it an epoch boundary?
+    if (maybeEpochStates.isPresent()) {
```
what about?

```java
    final Optional<StateAndBlockSummary> maybeEpochState =
        maybeEpochStates.flatMap(epochStates -> Optional.ofNullable(epochStates.get(blockRoot)));
    if (maybeEpochState.isPresent()) {
      return SafeFuture.completedFuture(maybeEpochState);
    }
```

rolfyone replied:
i really dont mind, can change to that format i guess.

rolfyone committed:
Refactoring, added trace in at trace level to assist in debugging the cache, and why a cache miss may occur.

This should hopefully address all review feedback, fairly substantial refactoring.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
add finals

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
cleanup

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
Merge remote-tracking branch 'upstream/master' into epoch-state-cache

rolfyone committed:
cleanup

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
Merge branch 'master' into epoch-state-cache

rolfyone committed:
Merge branch 'master' into epoch-state-cache

tbenr reviewed (approved):
LGTM
